sourcegraph/sourcegraph
# Add Self Serve Cody API client to dotcom

This PR is the first of many to fetch Cody's subscription data from SSC (Self-Serve Cody) into dotcom and integrate at the required places. This PR only takes care of the "fetching" part. 

Right now, we know if the user has opted for a pro trial by looking at the `users.cody_pro_enabled_at` column in the dotcom database. 

From Feb 1st, tentatively, we will be allowing users to put in their credit cards on SSC, which will create a `trialling` subscription for them on SSC. We will need the combined info of dotcom and SSC to support the UI.

After the Feb 15th release, SSC will be the only source of truth for the user's Cody subscription. Part of the code, added as part of this PR, will be removed after the release.

Cody Clients and Cody Management page on dotcom uses dotcom's API to know about the user's Cody subscription info. To support that, we need to fetch subscription info from SSC into dotcom.

**As part of this PR, we are:**
1. Introducing SSC client in `/internal/ssc`, which allows us to fetch the user's Cody Subscription from SSC.
2. Consolidating data from dotcom & SSC to figure out the user's subscription in `internal/cody/subscription`.
3. Make fetching info from SSC controlled by a feature flag: `use-ssc-for-cody-subscription`.
4. Updating `user.codyProEnabled` in graphql API to use `internal/cody/subscription`, which integrates it with subscription data from SSC.
5. Removing `user.codyProEnabledAt` from graphql API. It is a redundant field and not used by any Cody Clients or the web. Confirmed by this [search](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/*+codyProEnabledAt&patternType=standard&sm=1) and double checking locally.
6. Adding unit tests.

**Next steps as part of follow-up PRs:**
1. Expose subscription status info from the graphql API. 
2. Update rate-limit logic at various places to use `interna/cody/subscription` for the user's SSC subscription info rather than dotcom's DB.

The above 2 next steps will make the PLG flow functional from the dotcom side. As a follow-up, we will also work on creating a whole new actor source for Cody Gateway to reach SSC for rate limit info directly.

## Test plan

Unit tests are added as part of this PR.

**Manual Testing Plan**
1. Set up and run SAMS locally.
4. Integrate SAMS as an openidconnect auth provider with dotcom locally. 
5. Create a user using the SAMS auth provider.
6. Enable feature flag `use-ssc-for-cody-subscription` from site admin.
7. Make a graphql API [request](https://sourcegraph.test:3443/api/console#%7B%22query%22%3A%22%7B%5Cn%20%20currentUser%20%7B%5Cn%20%20%20%20codyProEnabled%5Cn%20%20%7D%5Cn%7D%5Cn%22%7D) to confirm `false`. 
8. Visit `localhost:9992/cody` and go through the subscription flow to sign up for Cody Pro.
9. Make a graphql API [request](https://sourcegraph.test:3443/api/console#%7B%22query%22%3A%22%7B%5Cn%20%20currentUser%20%7B%5Cn%20%20%20%20codyProEnabled%5Cn%20%20%7D%5Cn%7D%5Cn%22%7D) again to confirm `true`.
